engine/access/ingestion/engine.go

```diff
@@ -676,10 +697,86 @@ func (e *Engine) requestMissingCollections(ctx context.Context) error {
 	return nil
 }
 
+func (e *Engine) indexMissingCollectionsFromExecutionData(ctx context.Context, lastFullHeight uint64) error {
+	// these are not set if catchup from execution data is disabled
+	if e.executionDataHeight == nil || e.execDataCache == nil {
+		return nil
+	}
+
+	latestExecDataHeight, err := e.executionDataHeight()
+	if err != nil {
+		return fmt.Errorf("could not get latest execution data available: %w", err)
+	}
+
+	if lastFullHeight >= latestExecDataHeight {
+		return nil
+	}
+
+	lg := e.log.With().
+		Uint64("last_full_height", lastFullHeight).
+		Uint64("latest_exec_data_height", latestExecDataHeight).
+		Logger()
+
+	lg.Trace().Msg("indexing missing collections from execution data")
+
+	start := time.Now()
+	indexedCount := 0
+	for i := lastFullHeight + 1; i <= latestExecDataHeight; i++ {
+		// find missing collections for block at height i
+		missingColls, err := e.missingCollectionsAtHeight(i)
+		if err != nil {
+			return fmt.Errorf("could not get missing collections for height %d: %w", i, err)
+		}
+
+		// nothing to do
+		if len(missingColls) == 0 {
+			lg.Trace().
+				Uint64("height", i).
+				Msg("no missing collections for height")
+			continue
+		}
+
+		execData, err := e.execDataCache.ByHeight(ctx, i)
```

Comment:
The entire execution data also contains trie update, which could be very big. 

I wonder if possible execution data cache can return only the collection, it should be possible since we generate a cid for each field, right?

Reply:
we generate a CID for each chunk data, but those also include trie updates. We'll be integrating this into the indexer soon, so collections will get indexed along with the rest of the data without reading it from disk a second time.